src/app/games/quarto/QuartoRules.ts

```diff
@@ -129,41 +133,190 @@ interface LineInfos {
+            .map((coord: Coord) => new HorizontalVictoryPattern(coord));
+    }
+
+}
+
+class DescendingDiagonalVictoryPattern extends VictoryPattern {
+
+    private static readonly DESCENDING_DIAGONAL: CoordSet = new CoordSet([
+        new Coord(0, 0),
+        new Coord(1, 1),
+        new Coord(2, 2),
+        new Coord(3, 3),
+    ]);
+
+    public constructor(initialCoord: Coord) {
+        super(1,
+              DescendingDiagonalVictoryPattern.DESCENDING_DIAGONAL,
+              initialCoord);
+    }
+
+    public override getAllPatterns(state: QuartoState): DescendingDiagonalVictoryPattern[] {
+        const maxX: number = state.getWidth() - 4;
+        const maxY: number = state.getHeight() - 4;
+        return state
+            .getCoordsAndContents()
+            .map((coordAndContent: { coord: Coord; content: QuartoPiece}) => coordAndContent.coord)
+            .filter((coord: Coord) => coord.x <= maxX)
+            .filter((coord: Coord) => coord.y <= maxY)
+            .map((coord: Coord) => new DescendingDiagonalVictoryPattern(coord));
+    }
+
+}
+
+class AscendingDiagonalVictoryPattern extends VictoryPattern {
+
+    private static readonly ASCENDING_DIAGONAL: CoordSet = new CoordSet([
+        new Coord(0, 0),
+        new Coord(1, -1),
+        new Coord(2, -2),
+        new Coord(3, -3),
+    ]);
+
+    public constructor(initialCoord: Coord) {
+        super(1,
+              AscendingDiagonalVictoryPattern.ASCENDING_DIAGONAL,
+              initialCoord);
+    }
+
+    public override getAllPatterns(state: QuartoState): AscendingDiagonalVictoryPattern[] {
+        const maxX: number = state.getWidth() - 4;
+        const minY: number = 3;
+        const maxY: number = state.getHeight() - 1;
+        return state
+            .getCoordsAndContents()
+            .map((coordAndContent: { coord: Coord; content: QuartoPiece}) => coordAndContent.coord)
+            .filter((coord: Coord) => coord.x <= maxX)
+            .filter((coord: Coord) => minY <= coord.y && coord.y <= maxY)
+            .map((coord: Coord) => new AscendingDiagonalVictoryPattern(coord));
+    }
+
+}
+
+class SquareVictoryPattern extends VictoryPattern {
+
+    private static readonly SQUARE: CoordSet = new CoordSet([
+        new Coord(0, 0),
+        new Coord(0, 1),
+        new Coord(1, 0),
+        new Coord(1, 1),
+    ]);
+
+    public constructor(initialCoord: Coord) {
+        super(2,
+              SquareVictoryPattern.SQUARE,
+              initialCoord);
+    }
+
+    public override getAllPatterns(state: QuartoState): SquareVictoryPattern[] {
+        const maxX: number = state.getWidth() - 2;
+        const maxY: number = state.getHeight() - 2;
+        return state
+            .getCoordsAndContents()
+            .map((coordAndContent: { coord: Coord; content: QuartoPiece}) => coordAndContent.coord)
+            .filter((coord: Coord) => coord.x <= maxX)
+            .filter((coord: Coord) => coord.y <= maxY)
```

Comment:
once again, I am asking for you filter rewrite

Reply:
Hahaha d'accord mr. Sanders !